src/generators/notebook/set_collapsed_state.ts

```diff
@@ -0,0 +1,157 @@
+import { Panel } from '@phosphor/widgets';
+import { parseHeading } from '../../utils/parse_heading';
+import { isMarkdown } from '../../utils/is_markdown';
+import { isDOM } from '../../utils/is_dom';
+
+/**
+ * Collapses or expands ("un-collapses") a notebook cell by either hiding or displaying its section-defined sub-cells (i.e., cells which have lower precedence).
+ *
+ * @private
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ * @param state - collapsed state (`true` if collapse; `false` if expand)
+ */
+function setCollapsedState(
+  tracker: INotebookTracker,
+  cell: Cell,
+  state: boolean
+): void {
+  // Ensure a widget is currently focused...
+  if (tracker.currentWidget === null) {
+    return;
+  }
+  // Guard against attempting to collapse already hidden cells...
+  if (state) {
+    if (cell.isHidden) {
+      return;
+    }
+  }
+  // Guard against attempting to expand cells which we did not collapse or are already expanded...
+  else if (
+    cell.model.metadata.has('toc-nb-collapsed') === false ||
+    cell.model.metadata.get('toc-nb-collapsed') === false
+  ) {
+    // FIXME: we can end up here when attempting to expand a section generated by a code cell
+    return;
+  }
+  if (cell.model.type !== 'markdown') {
+    // FIXME: if we're here, it's because the user is attempting to collapse/expand a section generated by a code cell
+    return;
+  }
+  const heading = parseHeading(cell.model.value.text);
+
+  // Guard against attempting to (un-)collapse cells which are not "collapsible" (i.e., do not define sections)...
+  if (heading === null) {
+    return;
+  }
+  const level = heading.level;
+
+  const widgets = tracker.currentWidget.content.widgets;
+  const len = widgets.length;
+  const idx = widgets.indexOf(cell);
+
+  // Guard against an unrecognized "cell" argument...
+  if (idx === -1) {
+    return;
+  }
+  // Search for notebook cells which are semantically defined as sub-cells...
+  for (let i = idx + 1; i < len; i++) {
+    const w = widgets[i];
+
+    // Cells which are neither Markdown nor code cells can be readily collapsed/expanded...
+    if (w.model.type !== 'markdown' && w.model.type !== 'code') {
+      w.setHidden(state);
+      continue;
+    }
+    // Markdown cells are relatively straightforward, as we can determine whether to collapse/expand based on the level of the **first** encountered heading...
+    if (w.model.type === 'markdown') {
+      const h = parseHeading(w.model.value.text);
+
+      // Check if a widget is at the same or higher level...
+      if (h && h.level >= 0 && h.level <= level) {
+        // We've reached the end of the section...
+        break;
+      }
+      // Collapse/expand a sub-cell by setting the its `hidden` state:
+      w.setHidden(state);
+      continue;
+    }
+    // Code cells are more involved, as we need to analyze the outputs to check for generated Markdown/HTML...
+    const c = w as CodeCell;
```

Comment:
Yikes, I didn't know this was going to be so involved! Thanks for doing all the hard work here!

Reply:
Yeah, ;) code cell output is hard, involved, and entails considerable complexity, as discussed elsewhere. This provides enough support, IMO, for the majority of output cell use cases.